subprojects/core-api/src/main/java/org/gradle/api/NamedDomainObjectCollectionSchema.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api;
+
+import org.gradle.api.reflect.TypeOf;
+
+/**
+ * Schema of named domain object collections.
+ *
+ * @param <T> Schema of types contained in the collection.
+ * @since 4.10
+ * @see NamedDomainObjectCollection
+ */
+@Incubating
+public interface NamedDomainObjectCollectionSchema<T extends NamedDomainObjectCollectionSchema.NamedDomainObjectSchema> extends Iterable<T> {
```

Comment:
Why the type parameter? I don't think it adds any meaning.

Instead I would have an `elements` property that returns the schema for each element. This way we can later add more stuff about the container (to allow generation of specific factory methods, for example).

Something like:

```
    Iterable<? extends NamedDomainObjectSchema> getElements()
```

`ExtensionsSchema` can override this to provide an iterable of `ExensionSchema`.

Reply:
That's a better idea, done